File: daemon/internal/store/tenancy/approvals.go
package tenancy

import (
	"context"
	"errors"

	"github.com/dopejs/dope-agent/daemon/internal/audit"
	"github.com/dopejs/dope-agent/daemon/internal/policy"
	"github.com/dopejs/dope-agent/daemon/internal/store"
)

// Approvals is the tenant-aware accessor for the approvals + decisions
// pair. Pass A semantics mirror tenancy.Runtime.
type Approvals struct {
	store   *store.SQLiteStore
	emitter *audit.Emitter
}

func NewApprovals(s *store.SQLiteStore, emitter *audit.Emitter) *Approvals {
	return &Approvals{store: s, emitter: emitter}
}

func (a *Approvals) emit(ctx context.Context, surface, resourceKind string) {
	if a == nil || a.emitter == nil {
		return
	}
	_ = a.emitter.Emit(ctx, surface, resourceKind)
}

func (a *Approvals) ListApprovalsForTenant(ctx context.Context) ([]policy.Approval, error) {
	tenantID, err := Require(ctx)
	if err != nil {
		return nil, err
	}
	return a.store.ListApprovalsForTenantRaw(ctx, tenantID)
}

func (a *Approvals) UpsertApprovalForTenant(ctx context.Context, approval policy.Approval) error {
	tenantID, err := Require(ctx)
	if err != nil {
		return err
	}
	if err := a.store.UpsertApproval(ctx, approval); err != nil {
		return err
	}
	if err := a.store.BindRowTenant(ctx, "approvals", "approval_id", approval.ApprovalID, tenantID); err != nil {
		if errors.Is(err, store.ErrCrossTenantRow) {
			a.emit(ctx, "store:UpsertApprovalForTenant", "approval")
			return ErrCrossTenantWrite
		}
		return err
	}
	return nil
}

func (a *Approvals) ListDecisionsForTenant(ctx context.Context) ([]policy.Decision, error) {
	tenantID, err := Require(ctx)
	if err != nil {
		return nil, err
	}
	return a.store.ListDecisionsForTenantRaw(ctx, tenantID)
}

func (a *Approvals) UpsertDecisionForTenant(ctx context.Context, decision policy.Decision) error {
	tenantID, err := Require(ctx)
	if err != nil {
		return err
	}
	if err := a.store.UpsertDecision(ctx, decision); err != nil {
		return err
	}
	if err := a.store.BindRowTenant(ctx, "decisions", "decision_id", decision.DecisionID, tenantID); err != nil {
		if errors.Is(err, store.ErrCrossTenantRow) {
			a.emit(ctx, "store:UpsertDecisionForTenant", "decision")
			return ErrCrossTenantWrite
		}
		return err
	}
	return nil
}

File: crates/iam/tenancy/src/approvals.rs
//! Tenant-aware accessor for the approvals + decisions pair.
//! Port of daemon/internal/store/tenancy/approvals.go.

use crate::{emit_denial, require, TenancyError};

/// Tenant-aware accessor for approvals and decisions.
pub struct Approvals {
    store: crate::SQLiteStore,
    emitter: Option<dope_audit::Emitter>,
}

impl Approvals {
    #[must_use]
    pub fn new(store: crate::SQLiteStore, emitter: Option<dope_audit::Emitter>) -> Self {
        Approvals { store, emitter }
    }

    fn emit(&self, surface: &str, resource_kind: &str) {
        emit_denial(&self.emitter, surface, resource_kind);
    }

    pub fn list_approvals_for_tenant(&self) -> Result<Vec<dope_policy::Approval>, TenancyError> {
        let tenant_id = require()?;
        self.store.list_approvals_for_tenant_raw(&tenant_id).map_err(TenancyError::from)
    }

    pub fn upsert_approval_for_tenant(&self, approval: &dope_policy::Approval) -> Result<(), TenancyError> {
        let tenant_id = require()?;
        self.store.upsert_approval(approval).map_err(TenancyError::from)?;
        match self.store.bind_row_tenant("approvals", "approval_id", &approval.approval_id, &tenant_id) {
            Err(e) if crate::SQLiteStore::is_cross_tenant_row(&e) => {
                self.emit("store:UpsertApprovalForTenant", "approval");
                Err(TenancyError::CrossTenantWrite)
            }
            other => other.map_err(TenancyError::from),
        }
    }

    pub fn list_decisions_for_tenant(&self) -> Result<Vec<dope_policy::Decision>, TenancyError> {
        let tenant_id = require()?;
        self.store.list_decisions_for_tenant_raw(&tenant_id).map_err(TenancyError::from)
    }

    pub fn upsert_decision_for_tenant(&self, decision: &dope_policy::Decision) -> Result<(), TenancyError> {
        let tenant_id = require()?;
        self.store.upsert_decision(decision).map_err(TenancyError::from)?;
        match self.store.bind_row_tenant("decisions", "decision_id", &decision.decision_id, &tenant_id) {
            Err(e) if crate::SQLiteStore::is_cross_tenant_row(&e) => {
                self.emit("store:UpsertDecisionForTenant", "decision");
                Err(TenancyError::CrossTenantWrite)
            }
            other => other.map_err(TenancyError::from),
        }
    }
}
